src/Mod/Part/App/MeasureClient.cpp

```diff
@@ -0,0 +1,484 @@
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+
+#include <Mod/Part/PartGlobal.h>
+
+#include <string>
+
+#include <TopoDS.hxx>
+#include <TopoDS_Vertex.hxx>
+#include <TopAbs.hxx>
+#include <BRepTools.hxx>
+#include <BRep_Tool.hxx>
+#include <BRepGProp.hxx>
+#include <BRepAdaptor_Curve.hxx>
+#include <BRepAdaptor_Surface.hxx>
+#include <TopExp.hxx>
+#include <GProp_GProps.hxx>
+#include <ShapeAnalysis_Edge.hxx>
+#include <gp_Circ.hxx>
+#include <BRepBuilderAPI_Copy.hxx>
+
+#include <DatumFeature.h>
+#include <App/Application.h>
+#include <App/Document.h>
+#include <App/DocumentObject.h>
+#include <Base/Console.h>
+#include <Base/Matrix.h>
+#include <Base/Rotation.h>
+#include <Base/Vector3D.h>
+
+#include <Mod/Measure/App/MeasureAngle.h>
+#include <Mod/Measure/App/MeasureDistance.h>
+#include <Mod/Measure/App/MeasureLength.h>
+#include <Mod/Measure/App/MeasurePosition.h>
+#include <Mod/Measure/App/MeasureArea.h>
+#include <Mod/Measure/App/MeasureRadius.h>
+
+#include "VectorAdapter.h"
+#include "PartFeature.h"
+
+#include "MeasureClient.h"
+
+using namespace Part;
+
+// From: https://github.com/Celemation/FreeCAD/blob/joel_selection_summary_demo/src/Gui/SelectionSummary.cpp
+
+// Should work with edges and wires
+static float getLength(TopoDS_Shape& wire){
+    GProp_GProps gprops;
+    BRepGProp::LinearProperties(wire, gprops);
+    return gprops.Mass();
+}
+
+static float getFaceArea(TopoDS_Shape& face){
+    GProp_GProps gprops;
+    BRepGProp::SurfaceProperties(face, gprops);
+    return gprops.Mass();
+}
+
+static float getRadius(TopoDS_Shape& edge){
+    // gprops.Mass() would be the circumference (length) of the circle (arc)
+    if (edge.ShapeType() == TopAbs_EDGE) {
+        BRepAdaptor_Curve adapt(TopoDS::Edge(edge));
+        if (adapt.GetType() != GeomAbs_Circle) {
+            // TODO: not sure what the error handling here should be. nan? 0.0?
+            return 0.0;
+        }
+        gp_Circ circle = adapt.Circle();
+        return circle.Radius();
+    }
+    return 0.0;
+}
+
+TopoDS_Shape getLocatedShape(const std::string& objectName, const std::string& subName)
+{
+    App::DocumentObject* obj = App::GetApplication().getActiveDocument()->getObject(objectName.c_str());
```

Comment:
All the functions depend on the active document which reduces its reliability and flexibility a lot. FreeCAD is designed to have a MDI and thus all the functions still must work if their own document is not active.

In many cases it's not even checked that the accessed object exists and then a segfault occurs. As an example I used the radius handler pointing to the circular edge of a cone. When switching to another document after changing the radius of the cone a crash occurs:
```
Program received signal SIGSEGV, Segmentation fault.
#0  /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7ff58c71e520]
#1  0x7ff566c97cba in getLocatedShape(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) from /home/werner/Projects/build_worktree/build3/Mod/Part/Part.so+0x117
#2  0x7ff566c98ff8 in MeasureRadiusHandler(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*) from /home/werner/Projects/build_worktree/build3/Mod/Part/Part.so+0x88
#3  0x7ff566caa27d in std::shared_ptr<Part::MeasureRadiusInfo> std::__invoke_impl<std::shared_ptr<Part::MeasureRadiusInfo>, std::shared_ptr<Part::MeasureRadiusInfo> (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*>(std::__invoke_other, std::shared_ptr<Part::MeasureRadiusInfo> (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*&&) from /home/werner/Projects/build_worktree/build3/Mod/Part/Part.so+0x6b
#4  0x7ff566ca8e58 in std::enable_if<is_invocable_r_v<std::shared_ptr<Part::MeasureInfo>, std::shared_ptr<Part::MeasureRadiusInfo> (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*>, std::shared_ptr<Part::MeasureInfo> >::type std::__invoke_r<std::shared_ptr<Part::MeasureInfo>, std::shared_ptr<Part::MeasureRadiusInfo> (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*>(std::shared_ptr<Part::MeasureRadiusInfo> (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*&&) from /home/werner/Projects/build_worktree/build3/Mod/Part/Part.so+0x6d
#5  0x7ff566ca6d19 in std::_Function_handler<std::shared_ptr<Part::MeasureInfo> (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*), std::shared_ptr<Part::MeasureRadiusInfo> (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*&&) from /home/werner/Projects/build_worktree/build3/Mod/Part/Part.so+0x6d
#6  0x7ff566fcf233 in std::function<std::shared_ptr<Part::MeasureInfo> (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)>::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*) const from /home/werner/Projects/build_worktree/build3/Mod/Measure/Measure.so+0x79
#7  0x7ff566fd8b35 in Measure::MeasureRadius::getMeasureInfoFirst() const from /home/werner/Projects/build_worktree/build3/Mod/Measure/Measure.so+0x22f
#8  0x7ff566fd86fe in Measure::MeasureRadius::recalculateRadius() from /home/werner/Projects/build_worktree/build3/Mod/Measure/Measure.so+0x3e
#9  0x7ff566fd86b4 in Measure::MeasureRadius::execute() from /home/werner/Projects/build_worktree/build3/Mod/Measure/Measure.so+0x1c
```
Instead of the two strings (for object and sub-name) you should consider using SubObjectT or PropertyLinkT from App/DocumentObserver.h

Reply:
This has been addressed. There's one remaining place where this might be an issue, see: https://github.com/FreeCAD/FreeCAD/pull/9750/files#r1572142516